### PR TITLE
chore(yaml): add generic CSI operator yaml with root mount

### PR DESCRIPTION
**Pull request by prateekpandey14**
Signed-off-by: prateekpandey14 <prateek.pandey@mayadata.io>

As a container process whenever any other pods get Restart/Delete, container has to umount the specific mounted paths and has to be unlinked any reference mounts from the host as well.
Because the CSI Node daemonsets mount the `/:/host` (host root),  and the daemonset container can see the other mounts from the host as well.
So when any pod gets restart/delete all the other mount related info has to be propagated to the CSI node container, if not then pod will be stuck in `Terminated` state due to error "device or resource busy" while unlink the mounts 

But there is different behavior has been observed against the containerd versions installed in k8s clusters.

Issue for reference https://github.com/containerd/containerd/issues/3667

**Konvoy Clusters With MountPropagation needs to be set explicitly**
- CentOS: 7.8 version
- Containerd version: 1.3.4
 
So when any pod gets restart/delete all the mount related info are not getting  propagated to the CSI node container, unless we explicitly set the MountPropagation, otherwise  pod will be stuck in `Terminated` state due to error "device or resource busy" while unlink the mounts 

```
failed to get sandbox ip: check network namespace closed: remove netns: unlinkat /var/run/netns/cni-413cfa2b-b915-7561-cbf3-1e24f1093e85: device or resource busy
```

Below are the error outputs that show the mounts can be seen and holded by a CSI Node process:

```sh
[root@konvoy-node5 ~]# grep cni-413cfa2b-b915-7561-cbf3-1e24f1093e85 /proc/*/mountinfo
/proc/28724/mountinfo:1853 1813 0:3 / /host/run/netns/cni-413cfa2b-b915-7561-cbf3-1e24f1093e85 rw,nosuid,nodev,noexec,relatime - proc proc rw

$ ps -aux | grep 28724
root     28724  0.0  0.0 490676 25548 ?        Ssl  20:53   0:00 /usr/local/bin/cstor-csi-driver --nodeid=konvoy-node5.mayalabs.io --endpoint=unix:///plugin/csi.sock --url=https://openebs.io --plugin=node
root     31892  0.0  0.0 112712   960 pts/1    S+   20:56   0:00 grep --color=auto 28724
```

**Followed below steps to reproduce the issue:**
```yaml
          volumeMounts:
            - name: host-root
              mountPath: /host
            - name: host-test
              mountPath: /test
     volumes:
        - name: host-root
          hostPath:
            path: /
            type: Directory
        - name: host-test
          hostPath:
            path: /root/test
            type: DirectoryOrCreate
```

1: check the initial mounts inside the container
```
kubectl exec -it openebs-cstor-csi-node-twtj6 -c openebs-csi-plugin -n openebs bash
$ mount | grep pawan
/dev/mapper/centos-root on /pawan type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
```

2: Mounted the loop device in test/vol on Host
```
$ mkdir test/vol

$ sudo mount /dev/loop0 test/vol

$ mount | grep test
/dev/loop0 on /root/test/vol type ext4 (rw,relatime,seclabel,data=ordered)
 ```

3: Restart the pod to have updated mount paths
```
root@konvoy-master ~]# kubectl exec -it openebs-cstor-csi-node-4tdwb -c openebs-csi-plugin -n openebs bash
root@konvoy-node3:/# mount | grep test
/dev/mapper/centos-root on /test type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
/dev/loop0 on /test/vol type ext4 (rw,relatime,seclabel,data=ordered)
```

5: Unmount the device from HOST and Check the mounts

```
$ sudo umount /dev/loop0

$ grep test /proc/*/mountinfo
/proc/5874/mountinfo:1030 1009 253:0 /root/test /test rw,relatime - xfs /dev/mapper/centos-root rw,seclabel,attr2,inode64,noquota
/proc/5874/mountinfo:1031 1030 7:0 / /test/vol rw,relatime - ext4 /dev/loop0 rw,seclabel,data=ordered
/proc/5999/mountinfo:1030 1009 253:0 /root/test /test rw,relatime - xfs /dev/mapper/centos-root rw,seclabel,attr2,inode64,noquota
/proc/5999/mountinfo:1031 1030 7:0 / /test/vol rw,relatime - ext4 /dev/loop0 rw,seclabel,data=ordered 
```

7: Mounts exists inside the container ( which is expected behavior as MountPropagation not set, but in older containerd version list gets updated without MountPropagation set)
```
root@konvoy-node3:/# mount | grep test
/dev/mapper/centos-root on /test type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
/dev/loop0 on /test/vol type ext4 (rw,relatime,seclabel,data=ordered)
```

8. Try to remove the unmounted path ( in case of older containerd version we able to remove path successfully)
```
$ sudo rm -rf test/vol
rm: cannot remove 'test/vol': Device or resource busy
```

### Other Cluster without MountPropagation Set
- CentOS 7 with containerd version: 1.2.13
- Local Cluster containerd version: 1.2.13 

The issue was not reproducible in older version of containerd ( 1.2.13)  as the mounts are getting propagated inside the container once they get umounted from the Host, without explicitly set the MountPropagation. Due to this behavior, we were not hitting the above issue.

Mounted a test Directory inside the pod,  and followed the below steps to identify the behavior.  

1: mount list inside the csi-node container
```sh
root@centos-worker-1:/# mount | grep test
   /dev/sda2 on /test type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
```

2: Created a loop device in worker-1 and mount on /var/test
```   
   $ sudo mount /dev/loop0 /var/test
```

3: check the update mount list ( mount list getting updated without any MountPropagation set)
```
$ root@centos-worker-1:/# mount | grep test
  /dev/sda2 on /test type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
  /dev/loop0 on /host/var/test type ext4 (rw,relatime,seclabel,data=ordered)
```
4: umount the dir on worker-1
```
$ sudo umount /var/test
```

5: Mount removed from the list 
```
$ root@centos-worker-1:/# mount | grep test
  /dev/sda2 on /test type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
```

**Solution:**
As the above issue refers the unmount has not been propagated to the container, to solve this we need to share the mounts from host to container using "HostToContainer" propagation.

```yaml
          volumeMounts: 
            - name: host-root
              mountPath: /host
              mountPropagation: "HostToContainer"
```

More info about  can be found in [PR](https://github.com/openebs/cstor-operators/pull/119) about the different drivers and blogs advices or does mount the host root :

### CODE REVIEW

File: deploy/csi-operator.yaml [hunk 1]
```diff
@@ -962,8 +977,12 @@ spec:
               # needed so that any mounts setup inside this container are
               # propagated back to the host machine.
               mountPropagation: "Bidirectional"
-            - name: iscsiadm-bin
+            - name: host-root
+              mountPath: /host
+              mountPropagation: "HostToContainer"
+            - name: chroot-iscsiadm
               mountPath: /sbin/iscsiadm
+              subPath: iscsiadm
```
Comment (mittachaitu) [inline deploy/csi-operator.yaml:985]: Q: Why we need subpath?

Comment (prateekpandey14) [reply deploy/csi-operator.yaml:985]: because we don't want to mount as volume , i.e. to not change the path inside the container

Comment (prateekpandey14) [reply deploy/csi-operator.yaml:985]: This also ensures that any symlinks get resolved to inside the chroot environment.

File: deploy/csi-operator.yaml [hunk 2]
```diff
@@ -861,6 +855,27 @@ roleRef:
...
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: openebs-cstor-csi-iscsiadm
+  namespace: kube-system
+data:
+  iscsiadm: |
+    #!/bin/sh
+    if [ -x /host/sbin/iscsiadm ]; then
+      chroot /host /sbin/iscsiadm "$@"
```
Comment (pawanpraka1) [inline deploy/csi-operator.yaml:868]: if PATH env is set, then we don't need to check for each path to see where binary is present (the last else part covers that).

File: deploy/csi-operator.yaml [hunk 3]
```diff
@@ -981,7 +1000,12 @@ spec:
           hostPath:
             path: /var/lib/kubelet/
             type: Directory
-        - name: iscsiadm-bin
+        - name: chroot-iscsiadm
+          configMap:
+            defaultMode: 0755
```
Comment (pawanpraka1) [inline deploy/csi-operator.yaml:1005]: why not 0555? we just need read and execute permission, do we need write also?

File: deploy/csi-operator.yaml [hunk 4]
```diff
@@ -981,7 +1000,12 @@ spec:
...
+            name: openebs-cstor-csi-iscsiadm
+        - name: host-root
           hostPath:
-            path: /sbin/iscsiadm
-            type: File
+            path: /
```
Comment (pawanpraka1) [inline deploy/csi-operator.yaml:1009]: mounting the root not in readonly mode seems to be little risky and it is the responsibility of the application running inside the conaitner, that it is not messing up with the data and corrupting anything.

File: deploy/csi-operator.yaml (hunk 3 above)
Comment (mittachaitu) [reply deploy/csi-operator.yaml:1005]: I observed iscsiadm creating required files in `/etc/iscsi/nodes` and `/etc/iscsi/send_targets`. So I think it needs permission.

Comment (pawanpraka1) [reply deploy/csi-operator.yaml:1005]: @mittachaitu this is the iscsiadm script that will be mounted inside the conatiner via configmap. We just need read and execute permission for a script to execute.

Comment (mittachaitu) [reply deploy/csi-operator.yaml:1005]: Ah got it(We can change to 555).

File: deploy/csi-operator.yaml (hunk 2 above)
Comment (prateekpandey14) [reply deploy/csi-operator.yaml:868]: There can be different bin paths present based on the distribution, and even in some cases, if user installed and PATH env is not set then there is problem right?

File: deploy/csi-operator.yaml (hunk 3 above)
Comment (prateekpandey14) [reply deploy/csi-operator.yaml:1005]: Yes 555 should work fine ... let me try that once .

File: deploy/csi-operator.yaml [hunk 5]
```diff
@@ -861,6 +855,27 @@ roleRef:
   name: openebs-cstor-csi-registrar-role
   apiGroup: rbac.authorization.k8s.io
 
+---
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: openebs-cstor-csi-iscsiadm
+  namespace: kube-system
```
Comment (kmova) [inline deploy/csi-operator.yaml:863]: Change to openebs-namespace and set the minimum version to 1.17

File: deploy/csi-operator.yaml [hunk 6]
```diff
@@ -497,14 +491,29 @@ status:
...
+  name: cstor.csi.openebs.io
+spec:
+  # Supports persistent inline volumes.
+  volumeLifecycleModes:
+  - Persistent
+  # To determine at runtime which mode a volume uses, pod info and its
+  # "csi.storage.k8s.io/ephemeral" entry are needed.
+  podInfoOnMount: true
+  attachRequired: true
+---
```
Comment (prateekpandey14) [inline deploy/csi-operator.yaml:507]: @kmova added the CSIDriver volumeLifecycle Mode as well since we are setting up k8s 1.17.x as minimum version

Comment (akhilerm) [reply deploy/csi-operator.yaml:507]: @prateekpandey14  Does this mean the minimum version of k8s for openebs from 1.12 release should be 1.17?

File: deploy/csi-operator.yaml [hunk 7]
```diff
@@ -497,14 +491,29 @@ status:
...
+
 kind: ClusterRoleBinding
 apiVersion: rbac.authorization.k8s.io/v1
 metadata:
   name: openebs-cstor-csi-snapshotter-binding
 subjects:
   - kind: ServiceAccount
     name: openebs-cstor-csi-controller-sa
-    namespace: kube-system
+    namespace: openebs
```
Comment (akhilerm) [inline deploy/csi-operator.yaml:516]: question: if in openebs namespace, can the default service account for other openebs components be used here?

File: deploy/csi-operator.yaml (hunk 6 above)
Comment (prateekpandey14) [reply deploy/csi-operator.yaml:507]: Not for openebs overall, its specific to CSI/CSPC with all the feature supported like migration/upgrade etc.

File: deploy/csi-operator.yaml (hunk 7 above)
Comment (prateekpandey14) [reply deploy/csi-operator.yaml:516]: I think its good practice to have a specific service account with certain permission specific to components.

File: deploy/csi-operator.yaml (hunk 2 above)
Comment (akhilerm) [reply deploy/csi-operator.yaml:883]: question: `PATH` will be set on the host right? Not inside container. Or does the PATH change when we do `chroot`

**Comment by mittachaitu** [comment]
adding hold merge label --- just observed iscsi stale sessions even after deleting the deployment.

**Comment by prateekpandey14** [comment]
@kmova PR fixes that handles the iscsi stale session https://github.com/openebs/cstor-csi/pull/108

**Pull request merged**